src/sample_datasets.rs

```diff
@@ -0,0 +1,371 @@
+                (left, right)
+            } else {
+                (right, left)
+            };
+            if !self.cache.contains_key(&key) {
+                self.cache.insert(
+                    key,
+                    (self.function)(
+                        &self.data.row(left).into_dyn(),
+                        &self.data.row(right).into_dyn(),
+                    ),
+                );
+            }
+            *self.cache.get(&key).unwrap()
+        }
+    }
+
+    /// Compute the distances from `left` to all points in `right`.
+    #[allow(clippy::ptr_arg)]
+    fn distances_from(&self, left: Index, right: &Indices) -> Vec<U> {
+        right
+            .par_iter()
+            .map(|&r| self.distance(left, r))
+            .collect::<Vec<U>>()
+    }
+
+    /// Compute distances between all points in `left` and `right`.
+    #[allow(clippy::ptr_arg)]
+    fn distances_among(&self, left: &Indices, right: &Indices) -> Vec<Vec<U>> {
+        left.par_iter()
+            .map(|&l| self.distances_from(l, right))
+            .collect::<Vec<Vec<U>>>()
+    }
+
+    /// Compute the pairwise distance between all points in `indices`.
+    #[allow(clippy::ptr_arg)]
+    fn pairwise_distances(&self, indices: &Indices) -> Vec<Vec<U>> {
+        self.distances_among(indices, indices)
+    }
+
+    /// Clears the internal cache.
+    fn clear_cache(&self) {
+        self.cache.clear()
+    }
+
+    /// Returns the number of elements in the internal cache.
+    fn cache_size(&self) -> Option<usize> {
+        Some(self.cache.len())
+    }
+}
+
+pub struct Fasta {
+    pub metric: &'static str,
+    pub mmap: memmap2::Mmap,
+    pub shape: &'static [usize],
+    num_sequences: usize,
+    seq_len: usize,
+    offset: usize,
+    function: Metric<u8, u64>,
+    cache: DashMap<(Index, Index), u64>,
+}
+
+impl fmt::Debug for Fasta {
+    fn fmt(&self, f: &mut fmt::Formatter) -> result::Result<(), fmt::Error> {
+        f.debug_struct("Fasta Dataset")
+            .field("metric", &self.metric)
+            .field("offset", &self.offset)
+            .field("shape", &self.shape)
+            .finish()
+    }
+}
+
+impl Fasta {
+    pub fn new(
+        metric: &'static str,
+        path: PathBuf,
+        shape: &'static [usize],
+        offset: usize,
+    ) -> Result<Self, String> {
+        let file = match File::open(path) {
+            Ok(f) => f,
+            Err(_) => return Err("Could not open file".to_string()),
+        };
+        let mmap = unsafe { match Mmap::map(&file) {
+            Ok(f) => f,
+            Err(_) => return Err("Could not turn file into memmap".to_string()),
+        } };
+
+        Ok(Fasta {
+            metric,
+            mmap,
+            shape,
+            num_sequences: shape[0],
+            seq_len: shape[1],
+            offset,
+            function: metric_new(metric)?,
+            cache: DashMap::new(),
+        })
+    }
+}
+
+impl Dataset<u8, u64> for Fasta {
```

Comment:
Some more of these methods could be defaults in the trait but I need to implement this for a few other structs before I have a decent idea for how to get that done